<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Normalize AMP-AD Expression Files From Sage • rnaSeqNormalizer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Normalize AMP-AD Expression Files From Sage">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rnaSeqNormalizer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/basicIntro.html">Introduction: a simple demo</a>
    </li>
    <li>
      <a href="../articles/sage.html">Normalize AMP-AD Expression Files From Sage</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Normalize AMP-AD Expression Files From Sage</h1>
                        <h4 class="author">Paul Shannon</h4>
            
            <h4 class="date">2019-11-07</h4>
      
      
      <div class="hidden name"><code>sage.Rmd</code></div>

    </div>

    
    
<style>
.main-container { width: 1200px; max-width:2800px;}
</style>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>Sage provides RNA-seq expression files from the AMP-AD consortium. They appear to be in a standard format: tab-delimited, 4 summary lines, Ensembl identifiers in the first column.</p>
<p>This vignette shows how to - read a file into a data.frame - perform some minor transformations on that data.frame to make it ready for rnaSeqNormalizer - choose parameters and perform the desired normalization - do some visual exploration of the results</p>
</div>
<div id="load-the-libraries-we-need" class="section level1">
<h1 class="hasAnchor">
<a href="#load-the-libraries-we-need" class="anchor"></a>Load the libraries we need</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(rnaSeqNormalizer)
<span class="co">#&gt; Loading required package: org.Hs.eg.db</span>
<span class="co">#&gt; Loading required package: AnnotationDbi</span>
<span class="co">#&gt; Loading required package: stats4</span>
<span class="co">#&gt; Loading required package: BiocGenerics</span>
<span class="co">#&gt; Loading required package: parallel</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'BiocGenerics'</span>
<span class="co">#&gt; The following objects are masked from 'package:parallel':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ, clusterExport, clusterMap, parApply,</span>
<span class="co">#&gt;     parCapply, parLapply, parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     IQR, mad, sd, var, xtabs</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     anyDuplicated, append, as.data.frame, basename, cbind, colnames, dirname, do.call, duplicated, eval,</span>
<span class="co">#&gt;     evalq, Filter, Find, get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget, order,</span>
<span class="co">#&gt;     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span>
<span class="co">#&gt;     table, tapply, union, unique, unsplit, which, which.max, which.min</span>
<span class="co">#&gt; Loading required package: Biobase</span>
<span class="co">#&gt; Welcome to Bioconductor</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     Vignettes contain introductory material; view with 'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span>
<span class="co">#&gt; Loading required package: IRanges</span>
<span class="co">#&gt; Loading required package: S4Vectors</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'S4Vectors'</span>
<span class="co">#&gt; The following object is masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     expand.grid</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Loading required package: EnsDb.Hsapiens.v79</span>
<span class="co">#&gt; Loading required package: ensembldb</span>
<span class="co">#&gt; Loading required package: GenomicRanges</span>
<span class="co">#&gt; Loading required package: GenomeInfoDb</span>
<span class="co">#&gt; Loading required package: GenomicFeatures</span>
<span class="co">#&gt; Loading required package: AnnotationFilter</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'ensembldb'</span>
<span class="co">#&gt; The following object is masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter</span>
<span class="co">#&gt; Loading required package: DESeq</span>
<span class="co">#&gt; Loading required package: locfit</span>
<span class="co">#&gt; locfit 1.5-9.1    2013-03-22</span>
<span class="co">#&gt; Loading required package: lattice</span>
<span class="co">#&gt;     Welcome to 'DESeq'. For improved performance, usability and functionality, please consider migrating to</span>
<span class="co">#&gt;     'DESeq2'.</span></code></pre></div>
<p>Create an instance of the normalizer, supplying it with a small example expression matrix. We expect that it will have gene identifiers as row names and sample identifiers as column names.</p>
<p><em>Note</em> that the file used here, <em>ROSMAP_all_counts_matrix.txt</em> is very large, and is <em>not</em> included in the package. This example therefore, works verbatim only on my build system. To reproduce this chunk of code, you must provide your own filepath variable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filepath &lt;-<span class="st"> "~/github/rnaSeqNormalizer/inst/extdata/sage/ROSMAP_all_counts_matrix.txt"</span>
<span class="kw"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span>(filepath))
tbl.mt &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(filepath, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">as.is=</span><span class="ot">TRUE</span>, <span class="dt">header=</span><span class="ot">TRUE</span>, <span class="dt">nrow=</span><span class="op">-</span><span class="dv">1</span>)
tbl.mt &lt;-<span class="st"> </span>tbl.mt[<span class="op">-</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">1</span><span class="op">:</span><span class="dv">4</span>),]   <span class="co"># remove first four lines, which have summary statistics</span>
<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(tbl.mt)[<span class="dv">1</span>] &lt;-<span class="st"> "ensembl_id"</span>   <span class="co"># give the first column the name we expect</span>
tbl.mt<span class="op">$</span>ensembl_id &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">"</span><span class="ch">\\</span><span class="st">.[0-9]+"</span>, <span class="st">""</span>, tbl.mt<span class="op">$</span>ensembl_id)   <span class="co"># remove version suffixes from the ids</span></code></pre></div>
<p>We demonstrate two normalizations. The results are quite different. First the “log+scale” normalization proposed by Max Robinson and Michael Wainberg.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rnaSeqNormalizer-class.html">rnaSeqNormalizer</a></span>(tbl.mt, <span class="dt">algorithm=</span><span class="st">"log+scale"</span>, <span class="dt">duplicate.selection.statistic=</span><span class="st">"mean"</span>)
mtx.<span class="dv">1</span> &lt;-<span class="st">  </span><span class="kw"><a href="../reference/getNormalizedMatrix.html">getNormalizedMatrix</a></span>(x)
<span class="kw"><a href="https://rdrr.io/r/stats/fivenum.html">fivenum</a></span>(mtx.<span class="dv">1</span>)</code></pre></div>
<p>Now use a standard Bioconductor algorithm using the DESeq package, <em>vst</em> (variance stabilization transformation)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/rnaSeqNormalizer-class.html">rnaSeqNormalizer</a></span>(tbl.mt, <span class="dt">algorithm=</span><span class="st">"vst"</span>, <span class="dt">duplicate.selection.statistic=</span><span class="st">"sd"</span>)
mtx.<span class="dv">2</span> &lt;-<span class="st">  </span><span class="kw"><a href="../reference/getNormalizedMatrix.html">getNormalizedMatrix</a></span>(x)
<span class="co">#&gt; Warning in .vst.normalize(mtx): the vst algorithm needs a matrix of integer counts; converting...</span>
<span class="kw"><a href="https://rdrr.io/r/stats/fivenum.html">fivenum</a></span>(mtx.<span class="dv">2</span>)</code></pre></div>
<p>Compare these two techniques:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(mtx.<span class="dv">1</span>), <span class="dt">main=</span><span class="st">"log+scale"</span>)</code></pre></div>
<p><img src="sage_files/figure-html/hist1-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(mtx.<span class="dv">2</span>), <span class="dt">main=</span><span class="st">"vst"</span>)</code></pre></div>
<p><img src="sage_files/figure-html/hist2-1.png" width="700"></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.6.1 (2019-07-05)</span>
<span class="co">#&gt; Platform: x86_64-apple-darwin15.6.0 (64-bit)</span>
<span class="co">#&gt; Running under: macOS Mojave 10.14.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib</span>
<span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] rnaSeqNormalizer_1.0.12   DESeq_1.36.0              lattice_0.20-38           locfit_1.5-9.1           </span>
<span class="co">#&gt;  [5] EnsDb.Hsapiens.v79_2.99.0 ensembldb_2.8.1           AnnotationFilter_1.8.0    GenomicFeatures_1.36.4   </span>
<span class="co">#&gt;  [9] GenomicRanges_1.36.1      GenomeInfoDb_1.20.0       org.Hs.eg.db_3.8.2        AnnotationDbi_1.46.1     </span>
<span class="co">#&gt; [13] IRanges_2.18.3            S4Vectors_0.22.1          Biobase_2.44.0            BiocGenerics_0.30.0      </span>
<span class="co">#&gt; [17] BiocStyle_2.12.0         </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] httr_1.4.1                  splines_3.6.1               bit64_0.9-7                 assertthat_0.2.1           </span>
<span class="co">#&gt;  [5] BiocManager_1.30.9          blob_1.2.0                  GenomeInfoDbData_1.2.1      Rsamtools_2.0.3            </span>
<span class="co">#&gt;  [9] yaml_2.2.0                  progress_1.2.2              pillar_1.4.2                RSQLite_2.1.2              </span>
<span class="co">#&gt; [13] backports_1.1.5             digest_0.6.21               RColorBrewer_1.1-2          XVector_0.24.0             </span>
<span class="co">#&gt; [17] htmltools_0.4.0             Matrix_1.2-17               XML_3.98-1.20               pkgconfig_2.0.3            </span>
<span class="co">#&gt; [21] biomaRt_2.40.5              genefilter_1.66.0           bookdown_0.14               zlibbioc_1.30.0            </span>
<span class="co">#&gt; [25] xtable_1.8-4                BiocParallel_1.18.1         tibble_2.1.3                annotate_1.62.0            </span>
<span class="co">#&gt; [29] SummarizedExperiment_1.14.1 lazyeval_0.2.2              survival_2.44-1.1           magrittr_1.5               </span>
<span class="co">#&gt; [33] crayon_1.3.4                memoise_1.1.0               evaluate_0.14               fs_1.3.1                   </span>
<span class="co">#&gt; [37] MASS_7.3-51.4               tools_3.6.1                 prettyunits_1.0.2           hms_0.5.2                  </span>
<span class="co">#&gt; [41] matrixStats_0.55.0          stringr_1.4.0               DelayedArray_0.10.0         Biostrings_2.52.0          </span>
<span class="co">#&gt; [45] compiler_3.6.1              pkgdown_1.4.1               rlang_0.4.0                 grid_3.6.1                 </span>
<span class="co">#&gt; [49] RCurl_1.95-4.12             bitops_1.0-6                rmarkdown_1.16              DBI_1.0.0                  </span>
<span class="co">#&gt; [53] curl_4.2                    R6_2.4.0                    GenomicAlignments_1.20.1    knitr_1.25                 </span>
<span class="co">#&gt; [57] rtracklayer_1.44.4          bit_1.1-14                  zeallot_0.1.0               rprojroot_1.3-2            </span>
<span class="co">#&gt; [61] ProtGenerics_1.16.0         desc_1.2.0                  stringi_1.4.3               Rcpp_1.0.2                 </span>
<span class="co">#&gt; [65] geneplotter_1.62.0          vctrs_0.2.0                 xfun_0.10</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#load-the-libraries-we-need">Load the libraries we need</a></li>
      <li><a href="#session-info">Session Info</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Paul Shannon.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
